File: ViewPager2/app/src/main/java/com/kuanquan/viewpager2/BaseFragment.java
package com.kuanquan.viewpager2;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.IdRes;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

/**
 * BaseFragment，处理ViewModelProvider的初始化
 */
public abstract class BaseFragment extends Fragment implements View.OnClickListener{

    protected Context context;
    protected FragmentActivity activity;
    protected View view;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        context = this.getActivity();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
//        view = initLayout(inflater, container);
        view = inflater.inflate(getLayoutId(),null,false);
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initView();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initData(savedInstanceState);
    }

    protected void addOnClickListeners(View.OnClickListener listener, @IdRes int... ids) {
        if (ids != null) {
            for (@IdRes int id : ids) {
                view.findViewById(id).setOnClickListener(listener);
            }
        }
    }

    /**
     * bind layout resource file
     *
     * @return id of layout resource
     */
//    protected abstract View initLayout(LayoutInflater inflater, ViewGroup container);
    protected abstract int getLayoutId();

    /**
     * init all views and add events
     */
    protected abstract void initView();

    protected abstract void initData(Bundle savedInstanceState);

    /**
     * 移动到position位置
     *
     * @param mRecyclerView RecyclerView
     * @param position      位置对应的角标
     */
    protected void smoothMoveToPosition(RecyclerView mRecyclerView, final int position) {
        int firstItem = mRecyclerView.getChildLayoutPosition(mRecyclerView.getChildAt(0));
        int lastItem = mRecyclerView.getChildLayoutPosition(mRecyclerView.getChildAt(mRecyclerView.getChildCount() - 1));
        if (position < firstItem) {
            mRecyclerView.smoothScrollToPosition(position);
        } else if (position <= lastItem) {
            int movePosition = position - firstItem;
            if (movePosition >= 0 && movePosition < mRecyclerView.getChildCount()) {
                int top = mRecyclerView.getChildAt(movePosition).getTop();
                mRecyclerView.smoothScrollBy(0, top);
            }
        } else {
            mRecyclerView.smoothScrollToPosition(position);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        this.activity = null;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.activity = (FragmentActivity) context;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        this.activity = null;
    }
}

File: DoubleListRecyclerView/baselibrary/src/main/java/com/mouble/baselibrary/base/BaseFragment.kt
package com.mouble.baselibrary.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.annotation.IdRes
import androidx.fragment.app.Fragment
import com.mouble.baselibrary.util.EventCenter
import com.mouble.baselibrary.util.LogUtil
import org.greenrobot.eventbus.EventBus

abstract class BaseFragment: Fragment(), View.OnClickListener {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (isBindEventBusHere()) {
            EventBus.getDefault().register(this)
        }
    }

    abstract fun isBindEventBusHere(): Boolean
    abstract fun getLayoutId(): Int
    abstract fun initView()
    abstract fun initData(savedInstanceState: Bundle?)
    abstract fun dataObserver()

    protected fun showToast(msg: String) {
        Toast.makeText(activity?.applicationContext, msg, Toast.LENGTH_SHORT).show()
    }

    private var progressDialog: LoadDialog? = null
    fun showProgressDialog() {
        activity?.run {
            progressDialog = LoadDialog(this).apply {
                setCanceledOnTouchOutside(false)
                setCancelable(true)
            }
        }

        progressDialog?.run {
            try {
                if (isShowing)
                    dismiss()
                else
                    show()
            } catch (e: Exception) {
                LogUtil.e("异常 = ", e)
            }
        }
    }

    fun dismissProgressDialog() {
        try {
            progressDialog?.run {
                if (isShowing) dismiss()
            }
        } catch (e: Exception) {
            LogUtil.e("进度弹框取消异常 = ", e)
        }
    }

    fun postEventBus(type: String) {
        EventBus.getDefault().post(
            EventCenter<Any>(
                type
            )
        )
    }

    fun postEventBusSticky(type: String) {
        EventBus.getDefault().postSticky(
            EventCenter<Any>(
                type
            )
        )
    }

    fun postEventBusSticky(type: String, obj: Any) {
        EventBus.getDefault().postSticky(
            EventCenter(
                type,
                obj
            )
        )
    }

    fun postEventBus(type: String, obj: Any) {
        EventBus.getDefault().post(
            EventCenter(
                type,
                obj
            )
        )
    }

    override fun onDestroy() {
        super.onDestroy()
        if (isBindEventBusHere()) EventBus.getDefault().unregister(this)
        if (progressDialog != null) {
            dismissProgressDialog()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(getLayoutId(), null, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initData(savedInstanceState)
    }

    // TODO 型参数组 修饰符 vararg
    protected fun addOnClickListeners(listener: View.OnClickListener, @IdRes vararg ids: Int) {
        for (id in ids) {
            view?.findViewById<View>(id)?.setOnClickListener(listener)
        }
    }
}